extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,330 @@
+                    .value_or("");
+            auto match =
+                JsonGetField<std::string>(configuration, "match").value_or("");
+            if (request_path == "" || match == "") {
+              return false;
+            }
+            struct headerData header;
+            if (!JsonArrayIterate(
+                    configuration, "request_methods",
+                    [&](const json& methods) -> bool {
+                      auto method = JsonValueAs<std::string>(methods);
+                      if (method.second !=
+                          Wasm::Common::JsonParserResultDetail::OK) {
+                        LOG_WARN("unexpected method");
+                        return false;
+                      }
+                      header.request_methods.insert(
+                          std::string(method.first.value()));
+                      return true;
+                    })) {
+              LOG_WARN("Failed to parse configuration for request methods.");
+              return false;
+            }
+
+            if (!JsonArrayIterate(
+                    configuration, "credentials",
+                    [&](const json& credentials) -> bool {
+                      auto credential = JsonValueAs<std::string>(credentials);
+                      if (credential.second !=
+                          Wasm::Common::JsonParserResultDetail::OK) {
+                        LOG_WARN("unexpected credential");
+                        return false;
+                      }
+                      header.encoded_credentials.insert(
+                          Base64::encode(credential.first.value().data(),
+                                         credential.first.value().size()));
+                      return true;
+                    })) {
+              LOG_WARN("Failed to parse configuration for credentials.");
+              return false;
+            }
+            auto path_iter = basic_auth_configuration_.find(match);
+            if (path_iter != basic_auth_configuration_.end() &&
+                path_iter->second.find(request_path) !=
+                    basic_auth_configuration_[match].end()) {
+              for (auto& methods : header.request_methods) {
+                basic_auth_configuration_[match][request_path]
+                    .request_methods.insert(methods);
+              }
+
+              for (auto& creds : header.encoded_credentials) {
+                basic_auth_configuration_[match][request_path]
+                    .encoded_credentials.insert(creds);
+              }
+            } else {
+              basic_auth_configuration_[match][request_path] = header;
+            }
+            return true;
+          })) {
+    LOG_WARN("Failed to parse configuration information.");
+  }
+  return true;
+}
+
+FilterHeadersStatus PluginContext::credentialsCheck(
+    std::unordered_map<std::string, PluginRootContext::headerData>::mapped_type
```

Comment:
mapped_type just resolves to headerData, right?

Reply:
yes

Reply:
Why not just use PluginRootContext::headerData then? `const PluginRootContext::headerData&`

Reply:
done!